File: src/main/java/info/gianlucacosta/chronos/ide/IdeInput.java
/*§
  ===========================================================================
  Chronos IDE
  ===========================================================================
  Copyright (C) 2015 Gianluca Costa
  ===========================================================================
  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
  ===========================================================================
*/

package info.gianlucacosta.chronos.ide;

import info.gianlucacosta.chronos.interpreter.atoms.BooleanAtom;
import info.gianlucacosta.chronos.interpreter.atoms.StringAtom;
import info.gianlucacosta.chronos.interpreter.exceptions.FailedReadException;
import info.gianlucacosta.chronos.interpreter.io.BasicInput;
import javafx.application.Platform;
import javafx.scene.control.Alert;
import javafx.scene.control.ButtonBar;
import javafx.scene.control.ButtonType;
import javafx.scene.control.TextInputDialog;

import java.util.Optional;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicReference;

class IdeInput extends BasicInput {
    @Override
    public StringAtom readString(String prompt) {
        final AtomicReference<Optional<String>> inputBuffer = new AtomicReference<>();
        final AtomicBoolean simpleLock = new AtomicBoolean(false);

        Platform.runLater(() -> {
            TextInputDialog dialog = new TextInputDialog();
            dialog.setTitle(IdeInfo.getInstance().name());
            dialog.setHeaderText(prompt);


            Optional<String> inputString = dialog.showAndWait();
            inputBuffer.set(inputString);


            synchronized (simpleLock) {
                simpleLock.set(true);
                simpleLock.notifyAll();
            }

        });

        synchronized (simpleLock) {
            while (!simpleLock.get()) {
                try {
                    simpleLock.wait();
                } catch (InterruptedException ex) {
                    //Just do nothing
                }
            }
        }

        Optional<String> inputString = inputBuffer.get();

        if (inputString.isPresent()) {
            return new StringAtom(inputString.get());
        } else {
            throw new FailedReadException("Interrupted input");
        }
    }


    @Override
    public BooleanAtom readBoolean(String prompt) {
        final AtomicReference<Optional<Boolean>> inputBuffer = new AtomicReference<>(Optional.empty());
        final AtomicBoolean simpleLock = new AtomicBoolean(false);

        Platform.runLater(() -> {
            Alert alert = new Alert(Alert.AlertType.CONFIRMATION);
            alert.setTitle(IdeInfo.getInstance().name());
            alert.setHeaderText(prompt);

            ButtonType yesButton = new ButtonType("Yes");
            ButtonType noButton = new ButtonType("No");
            ButtonType cancelButton = new ButtonType("Cancel", ButtonBar.ButtonData.CANCEL_CLOSE);

            alert.getButtonTypes().setAll(yesButton, noButton, cancelButton);

            Optional<ButtonType> result = alert.showAndWait();
            if (result.get() == yesButton) {
                inputBuffer.set(Optional.of(true));
            } else if (result.get() == noButton) {
                inputBuffer.set(Optional.of(false));
            }

            synchronized (simpleLock) {
                simpleLock.set(true);
                simpleLock.notifyAll();
            }
        });

        synchronized (simpleLock) {
            while (!simpleLock.get()) {
                try {
                    simpleLock.wait();
                } catch (InterruptedException ex) {
                    //Just do nothing
                }
            }
        }

        Optional<Boolean> inputBoolean = inputBuffer.get();

        if (inputBoolean.isPresent()) {
            return new BooleanAtom(inputBoolean.get());
        } else {
            throw new FailedReadException("Interrupted input");
        }
    }
}

File: src/main/scala/info/gianlucacosta/chronos/ide/IdeInput.scala
/*§
  ===========================================================================
  Chronos IDE
  ===========================================================================
  Copyright (C) 2015-2016 Gianluca Costa
  ===========================================================================
  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
  ===========================================================================
*/

package info.gianlucacosta.chronos.ide


import java.util.concurrent.atomic.{AtomicBoolean, AtomicReference}

import info.gianlucacosta.chronos.chronos_ide.{ArtifactInfo => IdeInfo}
import info.gianlucacosta.chronos.interpreter.atoms.{BooleanAtom, StringAtom}
import info.gianlucacosta.chronos.interpreter.exceptions.FailedReadException
import info.gianlucacosta.chronos.interpreter.io.BasicInput

import scalafx.application.Platform
import scalafx.scene.control.{Alert, ButtonBar, ButtonType, TextInputDialog}


class IdeInput extends BasicInput {

  override def readString(prompt: String): StringAtom = {
    val inputBuffer = new AtomicReference[Option[String]]()
    val simpleLock = new AtomicBoolean(false)


    Platform.runLater {
      val dialog = new TextInputDialog {
        title = IdeInfo.name
        headerText = prompt
      }

      val inputString = dialog.showAndWait()
      inputBuffer.set(inputString)

      simpleLock.synchronized {
        simpleLock.set(true)
        simpleLock.notifyAll()
      }
    }


    simpleLock.synchronized {
      while (!simpleLock.get()) {
        try {
          simpleLock.wait()
        } catch {
          case ex: InterruptedException => //Just do nothing
        }
      }
    }

    val inputString = inputBuffer.get()

    if (inputString.nonEmpty) {
      new StringAtom(inputString.get)
    } else {
      throw new FailedReadException("Interrupted input")
    }
  }


  override def readBoolean(prompt: String): BooleanAtom = {
    val inputBuffer = new AtomicReference[Option[Boolean]](None)
    val simpleLock = new AtomicBoolean(false)

    Platform.runLater {
      val alert = new Alert(Alert.AlertType.Confirmation) {
        title = IdeInfo.name
        headerText = prompt
      }

      val yesButton = new ButtonType("Yes")
      val noButton = new ButtonType("No")
      val cancelButton = new ButtonType("Cancel", ButtonBar.ButtonData.CancelClose)

      alert.buttonTypes = Seq(
        yesButton,
        noButton,
        cancelButton
      )

      val result = alert.showAndWait()

      inputBuffer.set(
        result match {
          case Some(`yesButton`) => Some(true)
          case Some(`noButton`) => Some(false)
          case _ => None
        })

      simpleLock.synchronized {
        simpleLock.set(true)
        simpleLock.notifyAll()
      }
    }


    simpleLock.synchronized {
      while (!simpleLock.get()) {
        try {
          simpleLock.wait()
        } catch {
          case ex: InterruptedException => //Just do nothing
        }
      }
    }

    val inputBoolean = inputBuffer.get()

    if (inputBoolean.nonEmpty) {
      new BooleanAtom(inputBoolean.get)
    } else {
      throw new FailedReadException("Interrupted input")
    }
  }
}
